src/main/java/org/chameleoncloud/api/admin/UserGroupRolesResource.java

```diff
@@ -0,0 +1,320 @@
+	    return groupPermission.viewPermission(group);
+	case "manage":
+	    return groupPermission.managePermission(group);
+	case "view-members":
+	    return groupPermission.viewMembersPermission(group);
+	case "manage-members":
+	    return groupPermission.manageMembersPermission(group);
+	case "manage-membership":
+	    return groupPermission.manageMembershipPermission(group);
+	default:
+	    return null;
+	}
+    }
+
+    private Boolean isRootGroup(GroupModel group) {
+
+	// check if the group is a root group
+	// if a root group, check if fine grained permission is enabled
+	// enable fine grained admin permission if not
+
+	if (group.getParent() != null) {
+	    groupPermission.setPermissionsEnabled(group, false);
+	    return false;
+	}
+	if (!groupPermission.isPermissionsEnabled(group)) {
+	    groupPermission.setPermissionsEnabled(group, true);
+	}
+	return true;
+    }
+
+    private void grantPermission(String permissionName, GroupModel group, GroupModel subGroup, String policyName)
+	    throws RuntimeException {
+
+	// associate the group policy to the scope permissions
+	// if policy does not exist, create one lazily
+
+	Policy groupScopePolicy = getPolicyByPermissionName(permissionName, group);
+	if (groupScopePolicy == null) {
+	    throw new RuntimeException(
+		    String.format("Can not get policy from fine grained permission %s", permissionName));
+	}
+	String groupId = group.getId();
+	Set<String> associatedPolicies = groupScopePolicy.getAssociatedPolicies().stream()
+		.filter(p -> "group".equals(p.getType())).map(Policy::getName).collect(Collectors.toSet());
+	String associatedPolicyName = UserGroupRoles.formatPolicyName(policyName, groupId);
+	if (!associatedPolicies.contains(associatedPolicyName)) {
+	    // policy not associated
+	    // check if the associated policy exists
+	    ResourceServer rs = groupScopePolicy.getResourceServer();
+	    Policy associatedPolicy = policyStore.findByName(associatedPolicyName, rs.getId());
+	    if (associatedPolicy == null) {
+		// create a policy
+		GroupPolicyRepresentation rep = new GroupPolicyRepresentation();
+		rep.setName(associatedPolicyName);
+		rep.addGroup(subGroup.getId(), true);
+		associatedPolicy = policyStore.create(rep, groupScopePolicy.getResourceServer());
+	    }
+	    groupScopePolicy.addAssociatedPolicy(associatedPolicy);
+	}
+    }
+
+    private void grantMemberPermissions(GroupModel group) {
+	grantPermission("view", group, group, UserGroupRoles.MEMBER);
+	grantPermission("view-members", group, group, UserGroupRoles.MEMBER);
+    }
+
+    private void grantManagerPermissions(GroupModel group, GroupModel managerSubGroup) {
+	grantPermission("manage-membership", group, managerSubGroup, UserGroupRoles.MANAGER);
+    }
+
+    private void grantAdminPermissions(GroupModel group, GroupModel adminSubGroup) {
+	grantPermission("manage", group, adminSubGroup, UserGroupRoles.ADMIN);
+	grantPermission("manage-members", group, adminSubGroup, UserGroupRoles.ADMIN);
```

Comment:
What's the difference b/w manage-members and manage-memberships? And what does "manage" provide that is different?

Reply:
https://www.keycloak.org/docs/latest/server_admin/#group